Common/src/grid_movement_structure.cpp

```diff
@@ -8441,40 +8457,146 @@ void CFreeFormDefBox::SetSphe2Cart_CornerPoints(CConfig *config) {
   
 }
 
+
+void CFreeFormDefBox::SetCGNS(CGeometry *geometry, unsigned short iFFDBox, bool original) {
+#ifdef HAVE_CGNS
+
+  char FFDBox_filename[MAX_STRING_SIZE];
+  bool new_file;
+  unsigned short iDim, iDegree, jDegree, kDegree;
+  unsigned short outDim;
+  unsigned int pos;
+  char zonename[33];
+  int FFDBox_cgns_file;
+  int cell_dim, phys_dim;
+  int cgns_base, cgns_family, cgns_zone, cgns_err, dummy;
+  const char * basename;
+
+  /*--- FFD output is always 3D (even in 2D problems),
+   this is important for debuging ---*/
+  nDim = geometry->GetnDim();
+  cell_dim = nDim,
+  phys_dim = 3;
+
+  SPRINTF (FFDBox_filename, "ffd_boxes.cgns");
+
+  if ((original) && (iFFDBox == 0)) new_file = true;
+  else new_file = false;
+  
+  if (new_file) {
+    cgns_err = cg_open(FFDBox_filename, CG_MODE_WRITE, &FFDBox_cgns_file);
+    if (cgns_err) cg_error_print();
+    cgns_err = cg_descriptor_write("Title", "Visualization of the FFD boxes generated by SU2_DEF." );
+    if (cgns_err) cg_error_print();
+  }
+  else {
+    cgns_err = cg_open(FFDBox_filename, CG_MODE_MODIFY, &FFDBox_cgns_file);
+    if (cgns_err) cg_error_print();
+  }
+
+  if (original) {
+    basename = "Original_FFD";
+  }
+  else {
+    basename = "Deformed_FFD";
+  }
+
+  if (iFFDBox == 0){
+    cgns_err = cg_base_write(FFDBox_cgns_file, basename, cell_dim, phys_dim, &cgns_base);
+    if (cgns_err) cg_error_print();
+  }
+  cgns_err = cg_family_write(FFDBox_cgns_file, cgns_base, Tag.c_str(), &cgns_family);
+  if (cgns_err) cg_error_print();
+  
+  cgsize_t dims[9];
+  dims[0] = lDegree+1;
+  dims[1] = mDegree+1;
+  if (cell_dim == 3){
+     dims[2] = nDegree+1;
+  }
+  cgsize_t pointlen = 1;
+  for(int ii=0; ii<cell_dim; ii++){
+    dims[ii+cell_dim] = dims[ii] -1;
+    dims[ii+2*cell_dim] = 0;
+    pointlen *= dims[ii];
+  }
+
+  double *buffer = new double[pointlen];
```

Comment:
About the compilation error you are getting, in general in SU2 you should use the typedef "su2double" instead of double or float for floating point variables. The exception is when those variables are then passed to external libraries (which seems to be the case) in which case you should use "passivedouble". Then when assigning to those variables, and the left hand side is of type "su2double" you should enclose the expression in SU2_TYPE::GetValue(), i.e. buffer[pos] = SU2_TYPE::GetValue(Coord_Control_Points[iDegree][jDegree][kDegree][iDim]);